### PR TITLE
ollama setup and utility to run the models locally

### PR DESCRIPTION
**Pull request by guruswarupa**
# Pull Request

## Title
ollama setup and utility to run the models locally

- I do not have a machine that is capable of running this but im sure its gonna work well. If you can please lemme know your review on how well the script works :)

## Checklist
- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no errors/warnings/merge conflicts.

### CODE REVIEW

File: tabs/utils/ollama.sh [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            2) show_model_info ;;
+            3) create_model ;;
+            4) run_model ;;
+            5) remove_model;;
+            6) printf "%b\n" "${GREEN}Exiting...${RC}"; exit 0 ;;
+            *) printf "%b\n" "${RED}Invalid choice. Please try again.${RC}" ;;
+        esac
+
+        printf "%b\n" "${YELLOW}Press Enter to continue...${RC}"
+        read -r
```
Comment (adamperkowski) [inline tabs/utils/ollama.sh:186]:
```suggestion
        read
```

https://github.com/ChrisTitusTech/linutil/pull/416#discussion_r1760768186

File: tabs/utils/tab_data.toml [hunk 2]
```diff
@@ -12,6 +12,10 @@ script = "bluetooth-control.sh"
 name = "Numlock on Startup"
 script = "numlock.sh"
 
+[[data]]
+name = "Ollama Chatbot"
```
Comment (adamperkowski) [inline tabs/utils/tab_data.toml:16]:
```suggestion
name = "Ollama"
```

Maybe just `Ollama` would be better. Up to you.

File: tabs/utils/ollama.sh [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    fi
+}
+
+create_model() {
+    clear
+    list_models
+    printf "%b\n" "${YELLOW}Let's create a new model in Ollama!${RC}"
+
+    # Prompt for base model
+    printf "%b\n" "${YELLOW}Enter the base model (e.g., llama3.1):${RC}"
```
Comment (adamperkowski) [inline tabs/utils/ollama.sh:104]: We could link https://ollama.com/library here
or list available models like in `run_model()`.

File: tabs/utils/ollama.sh [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    read -r base_model
+
+    # Prompt for custom model name
+    printf "%b\n" "${YELLOW}Enter a name for the new customized model:${RC}"
+    read -r custom_model_name
+
+    # Prompt for temperature setting
+    printf "%b\n" "${YELLOW}Enter the desired temperature (higher values are more creative, lower values are more coherent, e.g., 1):${RC}"
+    read -r temperature
```
Comment (adamperkowski) [inline tabs/utils/ollama.sh:113]: Defaults here? Don't know how ollama's config files work but it seems you're calling an empty var (line 124) if user just hits ENTER.

**Review by adamperkowski — CHANGES_REQUESTED** [review]

File: tabs/utils/ollama.sh (hunk 3 above)
Comment (guruswarupa) [reply tabs/utils/ollama.sh:104]: ive run list_models fn above. check properly

**Comment by guruswarupa** [comment]
@adamperkowski ive fixed older issues. crosscheck once

File: tabs/utils/ollama.sh [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    printf "7. Gemma 2 - 9B (5.5GB)\n"
+    printf "8. Gemma 2 - 27B (16GB)\n"
+    printf "9. Mistral - 7B (4.1GB)\n"
+    printf "10. Moondream 2 - 1.4B (829MB)\n"
+    printf "11. Neural Chat - 7B (4.1GB)\n"
+    printf "12. Starling - 7B (4.1GB)\n"
+    printf "13. Code Llama - 7B (3.8GB)\n"
+    printf "14. Llama 2 Uncensored - 7B (3.8GB)\n"
+    printf "15. LLaVA - 7B (4.5GB)\n"
+    printf "16. Solar - 10.7B (6.1GB)\n"
```
Comment (adamperkowski) [inline tabs/utils/ollama.sh:53]: Move these to a separate function / string. Same here:
https://github.com/ChrisTitusTech/linutil/blob/7a469251d9cd5ca99c7587294e8bc5b91dc5e9b3/tabs/utils/ollama.sh#L95-L112
https://github.com/ChrisTitusTech/linutil/blob/7a469251d9cd5ca99c7587294e8bc5b91dc5e9b3/tabs/utils/ollama.sh#L68-L86
https://github.com/ChrisTitusTech/linutil/blob/7a469251d9cd5ca99c7587294e8bc5b91dc5e9b3/tabs/utils/ollama.sh#L118-L139

File: tabs/utils/ollama.sh [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+            2) show_model_info ;;
+            3) create_model ;;
+            4) run_model ;;
+            5) remove_model;;
+            6) printf "%b\n" "${GREEN}Exiting...${RC}"; exit 0 ;;
+            *) printf "%b\n" "${RED}Invalid choice. Please try again.${RC}" ;;
+        esac
+
+        printf "%b\n" "${YELLOW}Press Enter to continue...${RC}"
+        read
```
Comment (jeevithakannan2) [inline tabs/utils/ollama.sh:227]: You can't have an empty read without variable. It is not POSIX compliant
```suggestion
        read foo
```

**Review by jeevithakannan2 — CHANGES_REQUESTED** [review]

File: tabs/utils/ollama.sh [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    printf "11. Neural Chat - 7B (4.1GB)\n"
+    printf "12. Starling - 7B (4.1GB)\n"
+    printf "13. Code Llama - 7B (3.8GB)\n"
+    printf "14. Llama 2 Uncensored - 7B (3.8GB)\n"
+    printf "15. LLaVA - 7B (4.5GB)\n"
+    printf "16. Solar - 10.7B (6.1GB)\n"
+    printf "%b\n" "${YELLOW}Let's create a new model in Ollama!${RC}"
+
+    # Prompt for base model
+    printf "%b\n" "${YELLOW}Enter the base model (e.g., llama3.1):${RC}"
```
Comment (adamperkowski) [inline tabs/utils/ollama.sh:115]:
```suggestion
    printf "%b\n" "${YELLOW}Enter the base model (e.g. '13' for codellama):${RC}"
```

File: tabs/utils/ollama.sh [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            2) show_model_info ;;
+            3) create_model ;;
+            4) run_model ;;
+            5) remove_model;;
+            6) printf "%b\n" "${GREEN}Exiting...${RC}"; exit 0 ;;
+            *) printf "%b\n" "${RED}Invalid choice. Please try again.${RC}" ;;
+        esac
+
+        printf "%b\n" "${YELLOW}Press Enter to continue...${RC}"
+        read dummy
```
Comment (adamperkowski) [inline tabs/utils/ollama.sh:227]:
```suggestion
        read -r dummy
```

Was wrong about this. Sorry.

**Review by adamperkowski — CHANGES_REQUESTED** [review]

File: tabs/utils/ollama.sh (hunk 8 above)
Comment (guruswarupa) [reply tabs/utils/ollama.sh:227]: :)

**Review by adamperkowski — APPROVED** [review]
Works like a charm :)

**Comment by guruswarupa** [comment]
> Works like a charm :)

hehe :) happy to hear that! can't wait for it to be pushed to main...

**Review by ChrisTitusTech — APPROVED** [review]

**Pull request merged**